src/Qir/Runtime/lib/QIR/delegated.cpp

```diff
@@ -25,27 +25,61 @@ static std::unordered_map<RESULT*, int>& AllocatedResults()
     return allocatedResults;
 }
 
+static Microsoft::Quantum::IRestrictedAreaManagement* RestrictedAreaManagement()
+{
+    return dynamic_cast<Microsoft::Quantum::IRestrictedAreaManagement*>(
+        Microsoft::Quantum::GlobalContext()->GetDriver());
+}
+
 extern "C"
 {
     Result quantum__rt__result_get_zero()
     {
         return Microsoft::Quantum::GlobalContext()->GetDriver()->UseZero();
     }
+
     Result quantum__rt__result_get_one()
     {
         return Microsoft::Quantum::GlobalContext()->GetDriver()->UseOne();
     }
 
-    QUBIT* quantum__rt__qubit_allocate() // NOLINT
+    QUBIT* quantum__rt__qubit_allocate()
     {
         return Microsoft::Quantum::GlobalContext()->GetDriver()->AllocateQubit();
     }
 
-    void quantum__rt__qubit_release(QUBIT* qubit) // NOLINT
+    void quantum__rt__qubit_release(QUBIT* qubit)
     {
         Microsoft::Quantum::GlobalContext()->GetDriver()->ReleaseQubit(qubit);
     }
 
+    QUBIT* quantum__rt__qubit_borrow()
+    {
+        // Currently we implement borrowing as allocation.
+        return quantum__rt__qubit_allocate();
+    }
+
+    void quantum__rt__qubit_return(QUBIT* qubit)
+    {
+        // Currently we implement borrowing as allocation.
+        quantum__rt__qubit_release(qubit);
+    }

```

Comment:
I'm a little confused as to why these are implemented here this way... the QIR spec doesn't describe borrowing, our QIR bridge has those functions commented out (they should probably be removed), and QirGeneration rewrite step in the compiler automatically turns any borrows into allocates. Since there are no callers for these and there is no infrastructure to generate calls to these and they aren't in the spec, why include them here as forwarders to allocate?

Reply:
Great question. From our last discussion with @bettinaheim I got under impression that eventually we would like to implement true borrowing. Forwarding in the rewrite step seems useful when runtime doesn't support it. But for the true borrowing implementation we need to support it in the runtime. First, we will forward on the level C interface. Then we will forward in qubit manager and simulators, and finally we will implement true support in qubit manager. I believe this is the plan at the moment.